rust/template/server.rs

```diff
@@ -0,0 +1,129 @@
+use differential_datalog::program::{RelId, Update, Response};
+use differential_datalog::record::{Record, UpdCmd, RelIdentifier};
+
+use api::*;
+use channel::*;
+
+use std::collections::{HashSet, HashMap};
+use std::sync::{Arc, Mutex};
+use std::fmt::{Debug, Formatter};
+use std::fmt;
+
+pub struct UpdatesSubscription<'a> {
+    observer: &'a mut Option<Arc<Mutex<dyn Observer<Update<super::Value>, String>>>>,
+}
+
+impl <'a> UpdatesSubscription<'a> {
+    pub fn new(observer: &'a mut Option<Arc<Mutex<dyn Observer<Update<super::Value>, String>>>>) -> Self {
+        UpdatesSubscription{observer: observer}
+    }
+}
+
+impl <'a> Subscription<'a> for UpdatesSubscription<'a> {
+    fn unsubscribe(&'a mut self) {
+        *self.observer = None;
+    }
+}
+
+pub struct DDlogServer
+{
+    prog: HDDlog,
+    outlets: Vec<Outlet>,
+    redirect: HashMap<RelId, RelId>
+}
+
+impl Debug for DDlogServer {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "DDlogServer")
+    }
+}
+
+impl DDlogServer
+{
+    pub fn new(prog: HDDlog, redirect: HashMap<RelId, RelId>) -> Self {
+        DDlogServer{prog: prog, outlets: Vec::new(), redirect: redirect}
+    }
+
+    pub fn stream(&mut self, tables: HashSet<RelId>) -> &mut Outlet {
+        let outlet = Outlet{tables : tables, observer : None};
+        self.outlets.push(outlet);
+        self.outlets.last_mut().unwrap()
+    }
+}
+
+pub struct Outlet
+{
+    tables: HashSet<RelId>,
+    observer: Option<Arc<Mutex<dyn Observer<Update<super::Value>, String>>>>
+}
+
+impl Observable<Update<super::Value>, String> for Outlet
+{
+    fn subscribe<'a>(&'a mut self,
+                     observer: Arc<Mutex<dyn Observer<Update<super::Value>, String>>>)
+                     -> Box<dyn Subscription + 'a>
+    {
+        self.observer = Some(observer);
+        Box::new(UpdatesSubscription::new(&mut self.observer))
+    }
+}
+
+impl Observer<Update<super::Value>, String> for DDlogServer
+{
+    fn on_start(&mut self) -> Response<()> {
+        self.prog.transaction_start()
+    }
+
+    fn on_commit(&mut self) -> Response<()> {
+        let changes = self.prog.transaction_commit_dump_changes()?;
+        for change in changes.as_ref().iter() {
+            println!{"Got {:?}", change};
+        }
+        for outlet in &mut self.outlets {
+            if let Some(ref observer) = outlet.observer {
+                let upds = outlet.tables.iter().flat_map(|table| {
+                    changes.as_ref().get(table).unwrap().iter().map(move |(val, weight)| {
```

Comment:
I can change `add_stream` to validate the id's.

Reply:
I guess it's still unsafe even if we validate the id's to exist in the table in `add_stream`, since some of the streamed tables might not have changed.

Reply:
I changed `add_stream` to validate the id's. When the id exists in the db but not in the delta, I ignore them when passing along the delta.

Reply:
Hmm there isn't actually a way to validate the ID's unless `do_store` is set to `true`. So for now I just ignore invalid ID's.